src/NHibernate/Cache/ReadWriteCache.cs

```diff
@@ -96,6 +98,7 @@ private int NextLockId()
 		/// </remarks>
 		public object Get(CacheKey key, long txTimestamp)
 		{
+			CheckCache();
 			using (_asyncReaderWriterLock.ReadLock())
```

Comment:
In the (strict) read-write strategy case, most cache operations acquire this lock. This class `Destroy` does dispose it, so most operations, in case of continued usage, would likely throw a disposed object exception from within the lock.

Somehow it is actually better than a NRE, but still not good: the strategies classes are "destroyed", so throwing a disposed object exception can be considered suitable. Excepted that it should be the strategy class throwing it directly, instead of some internal object it holds. Letting this been thrown by some internal object can puzzle end users as whether it is a bug inside NHibernate or not, like that strategy class having a bug causing usage of a disposed object.

Reply:
Now it is protected by acquiring the cache prior to accessing the lock, with the cache acquisition throwing if the strategy is already destroyed.
This is a bit brittle pattern since we could acquire the cache later...